### ISSUE TITLE
How to make ForceScheduler show available branches in droplist ?

### ISSUE THREAD

**Issue by jakobhu** [buildbot/buildbot#6088]
I'm using a **ForceScheduler** to allow users to schedule builds using a button from our local buildbot webpage.

I would like the UI to show a droplist with the available branches for a particular repository. This way users don't have to type their feature branch names correctly each time.  How do I achieve that ?

So far I've looked at exploiting the **codebases** option of **ForceScheduler** (see example http://docs.buildbot.net/current/manual/configuration/schedulers.html#forcescheduler-scheduler) - but the **branch** choices are hardcoded in all the examples I've seen. I'd like to avoid updating the buildbot config whenever a users pushes a new branch in the tracked repos.

**Comment by heytred** [comment]
What I've done is with a CodebaseParameter with a ChoiceStringParameter for the branch value. 

http://docs.buildbot.net/latest/manual/configuration/schedulers.html#codebaseparameter
http://docs.buildbot.net/latest/manual/configuration/schedulers.html?highlight=choicestringparameter#choicestringparameter

Example might look something roughly like this?

`
from buildbot.plugins import util

PROJECT = "proj_name"
REPO = "repo_name"
BRANCHES = [
    "main",
    "dev"
]

project = util.FixedParameter(name="project", default=PROJECT)
repo = util.FixedParameter(name="repository", default=REPO)
branch = util.ChoiceStringParameter(
    name="branch",
    choices=BRANCHES,
    default=BRANCHES[0],
    strict=True,
    multiple=False
)

codebase = util.CodebaseParameter(
        "",
        branch=branch,
        project=project,
        repository=repo,
        )

sched = schedulers.ForceScheduler(name="force", builderNames=["test"], codebases=[codebase])
`

**Comment by jakobhu** [comment]
Yes something like that, however; the branch-names are still hard-coded in that example (i.e. "main", "dev").
I'm looking for examples or ideas on how to make buildbot discover the branch names of a given repos.

Could I have to have an external service monitoring the repository, storing the discovered branches in a file - then load them on the buildbot master to construct the list of choices ? I'm not sure if that's feasible or even possible with buildbot.

**Comment by heytred** [comment]
Take a look at the poller for the VCS you are using (for example, link to GitPoller: http://docs.buildbot.net/latest/manual/configuration/changesources.html?highlight=change%20source%20poller#gitpoller). 

You'd probably be able to use the `branches` parameter set to `True` for all branches and create a subclass that during config then caches the branch names somewhere for the master to reference. I think there's an existing solution or service in Buildbot.
